### DISCUSSION TITLE
using filters and SELECT BOX

### DISCUSSION THREAD

**Discussion by rogerad** [rappasoft/laravel-livewire-tables#361] [General]
hi  Anthony 

filters  WORK EG  ```

 'categories' => Filter::make('categories')
                ->select([
                    '' => 'All',
                    'uncategorized' => 'uncategorized',
                    //'no' => 'No',
                ]),

BUT I CANNOT ADD 100 'S OF OPTIONS dynamically AT run Time

how do i do it because i will not know the categories until runtime when program is running. The categories are stored in another table and must be loaded at runtime
```

**Comment by rappasoft** [comment]
```php
public function filters(): array
{
  // Should probably get from cache so you don't hit the database on each render
  $myOptions = Model::where('this', 'that')->get()->toArray();

  return [
      'options' => Filter::make('Active')
          ->select(
            array_merge([
              '' => 'Any',
              'yes' => 'Yes',
              'no' => 'No',
            ], $myOptions)
          ),
  ];
}
```

**Reply by rogerad** [reply]
hi  i tried your suggestion.
 
<img width="566" alt="rapas" src="https://user-images.githubusercontent.com/18638656/122853735-09344200-d356-11eb-8ef9-9117cbbbaa32.png">
now the select box is correct but option is NOT found when i search for it with the filter selection
```
public function filters(): array
    {
        $c1ategories = Store_1_c1ategorie:: pluck('name')->sortBy('name')->toArray();
        return [
          

            'options' => Filter::make('categories')
                ->select(
                    array_merge([
                        '' => 'Any',
                       
                        'no' => 'No',
                    ], $c1ategories)
                ),
        ];
```

**Reply by rappasoft** [reply]
... Did you do this part? https://github.com/rappasoft/laravel-livewire-tables/wiki/Creating-filters#2-specifying-how-that-filter-acts-on-the-query

**Reply by rogerad** [reply]
yes   ```
  public
    function query(): Builder
    {
        return Product::query()
            ->when($this->getFilter('categories'), fn($query, $categories) => $query->where('categories', $categories));
        //  ->when($this->getFilter('active'), fn ($query, $active) => $query->where('active', $active === 'yes'));
    }
```

**Reply by rappasoft** [reply]
Your filter name is `options` not `categories`.

**Reply by rogerad** [reply]
hi
 when i changed options to categories **nothing came up** when i searched with the filter 
 please look at picture
<img width="644" alt="rapas 2" src="https://user-images.githubusercontent.com/18638656/122854931-ca06f080-d357-11eb-8c67-c2a0fa08e18b.png">

**Comment by rappasoft** [comment]
If categories is a one to many, you have to check the relationship. Something like:

```php
->when($this->getFilter('options'), fn($query, $option) => $query->whereHas('categories', function($query) use($option) {
    return $query->where('name', $option);
});
```

Or something like that.

You have to debug until you find where your query is wrong.

**Reply by rogerad** [reply]
ok it is now working the problem was the id was being stored with the name in the select box as soon as i plucked the name twice  using your first suggestion 
everything started working
in case anyone else cannot get a select box working they must pluck twice eg

`$c1ategories = Store_1_c1ategorie:: pluck('name', 'name')->sortBy('name')->toArray();`
